ecc/bls12-377/fr/poseidon2/gkrgates/gkrgates.go

```diff
@@ -0,0 +1,193 @@
+)
+
+// The GKR gates needed for proving Poseidon2 permutations
+
+// extKeySBoxGate applies the external matrix mul, then adds the round key, then applies the sBox
+// because of its symmetry, we don't need to define distinct x1 and x2 versions of it
+type extKeySBoxGate struct {
+	roundKey fr.Element
+}
+
+func (g *extKeySBoxGate) Evaluate(x ...fr.Element) fr.Element {
+	if len(x) != 2 {
+		panic("expected 2 inputs")
+	}
+
+	x[0].
+		Double(&x[0]).
+		Add(&x[0], &x[1]).
+		Add(&x[0], &g.roundKey)
+	return sBox2(x[0])
+}
+
+func (g *extKeySBoxGate) Degree() int {
+	return poseidon2.DegreeSBox()
+}
+
+// for x1, the partial round gates are identical to full round gates
+// for x2, the partial round gates are just a linear combination
+// TODO @Tabaie eliminate the x2 partial round gates and have the x1 gates depend on i - rf/2 or so previous x1's
+
+// extGate2 applies the external matrix mul, outputting the second element of the result
+type extGate2 struct{}
+
+func (extGate2) Evaluate(x ...fr.Element) fr.Element {
+	if len(x) != 2 {
+		panic("expected 2 inputs")
+	}
+	x[1].
+		Double(&x[1]).
+		Add(&x[1], &x[0])
+	return x[1]
+}
+
+func (g extGate2) Degree() int {
+	return 1
+}
+
+// intGate2 applies the internal matrix mul, returning the second element
+type intGate2 struct {
+}
+
+func (g intGate2) Evaluate(x ...fr.Element) fr.Element {
+	if len(x) != 2 {
+		panic("expected 2 inputs")
+	}
+	x[0].Add(&x[0], &x[1])
+	x[1].
+		Double(&x[1]).
+		Add(&x[1], &x[0])
+	return x[1]
+}
+
+func (g intGate2) Degree() int {
+	return 1
+}
+
+// intKeySBoxGateFr applies the second row of internal matrix mul, then adds the round key, then applies the sBox
+type intKeySBoxGate2 struct {
+	roundKey fr.Element
+}
+
+func (g *intKeySBoxGate2) Evaluate(x ...fr.Element) fr.Element {
+	if len(x) != 2 {
+		panic("expected 2 inputs")
+	}
+	x[0].Add(&x[0], &x[1])
+	x[1].
+		Double(&x[1]).
+		Add(&x[1], &x[0]).
+		Add(&x[1], &g.roundKey)
+
+	return sBox2(x[1])
+}
+
+func (g *intKeySBoxGate2) Degree() int {
+	return poseidon2.DegreeSBox()
+}
+
+type extGate struct{}
+
+func (g extGate) Evaluate(x ...fr.Element) fr.Element {
+	if len(x) != 2 {
+		panic("expected 2 inputs")
+	}
+	x[0].
+		Double(&x[0]).
+		Add(&x[0], &x[1])
+	return x[0]
+}
+
+func (g extGate) Degree() int {
+	return 1
+}
+
+// sBox2 is Hash.sBox for t=2
+func sBox2(x fr.Element) fr.Element {
+	var y fr.Element
+	y.Square(&x).Square(&y).Square(&y).Square(&y).Mul(&x, &y)
+	return y
+}
+
+var initOnce sync.Once
+
+// RegisterGkrGates registers the Poseidon2 permutation gates for GKR
+func RegisterGkrGates() {
+	initOnce.Do(
+		func() {
+			p := poseidon2.NewDefaultParameters()
+			halfRf := p.NbFullRounds / 2
+
+			gateNameX := func(i int) string {
+				return fmt.Sprintf("x-round=%d%s", i, p.String())
+			}
+			gateNameY := func(i int) string {
+				return fmt.Sprintf("y-round=%d%s", i, p.String())
+			}
+
+			fullRound := func(i int) {
+				gkr.Gates[gateNameX(i)] = &extKeySBoxGate{
```

Comment:
I copy-pasted this part - but here you have unsynchronized access to `gkr.Gates`. It may lead to situations where several places the map is updated at the same time, leading to only the last write to succeed.

Better approach is to use the registry approach where `gkr` package provides function `RegisterGate` and performs assignment in synchronous manner.

Reply:
Agreed. Will address it in the next PR.